<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Module 4 – Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../agenda.html"> 
<span class="menu-text">Agenda</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-training-materials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Training materials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-training-materials">    
        <li>
    <a class="dropdown-item" href="../Day1/QC.html">
 <span class="dropdown-text">QC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day1/Assembly.html">
 <span class="dropdown-text">Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/Annotation.html">
 <span class="dropdown-text">Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/AMR-detection.html">
 <span class="dropdown-text">AMR-detection</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Phylogenetics.html">
 <span class="dropdown-text">Phylogenetics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Comparative-Genomics.html">
 <span class="dropdown-text">Comparative-Genomics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Assessment/assessment.html">
 <span class="dropdown-text">Assessment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Module 4</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Module 4</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>SARS-CoV-2 surveillance in Romania</p>
<ul>
<li><p>This exercise used tools developed by The Centre for Genomic Pathogen Surveillance (CGPS) at University of Oxford</p></li>
<li><p>Similar exercises have been developed as part of T3Connect <a href="https://wcscourses.github.io/T3connectResources/" class="uri">https://wcscourses.github.io/T3connectResources/</a></p></li>
</ul>
<p>Genomic surveillance of pathogens aims to understand the emergence and dissemination of pathogens or their lineages of risk with the ultimate goal of implementing evidence-based interventions to protect public health. Epidemiological data from patients is collected by healthcare professionals. Data on species identification and any phenotypic or molecular characterization of the isolates is often generated by the microbiology laboratories linked to healthcare facilities and/or the reference laboratory. In an ideal scenario, the different sources of laboratory data are stored in a centralised surveillance system and database (such as <a href="https://whonet.org/">WHONET</a>). However, these systems rarely also incorporate genomic data produced by bioinformaticians.&nbsp;</p>
<p>The Centre for Genomic Pathogen Surveillance develops free web applications for integration and visualisation of surveillance (and other) data, called <a href="https://data-flo.io/">data-flo</a> and <a href="http://www.microreact.org">Microreact</a>. The aim of this module is to highlight the role of the various analytics and sources of data in pathogen surveillance, and how streamlined data integration is essential for real-time decision making, while also introducing resources that can facilitate the inclusion of this topic in teaching curricula.</p>
<p><br>
</p>
<section id="learning-outcomes" class="level1">
<h1>Learning Outcomes</h1>
<p>At the end of this module participants will be able to:</p>
<ol type="1">
<li><p>Distinguish the different components of genomic surveillance and their sources</p></li>
<li><p>Identify analytic tools to extract from genomes useful information for surveillance&nbsp;</p></li>
<li><p>Utilise data-flo to join different sources of data</p></li>
<li><p>Utilise Microreact to recognize the interplay between different types of data</p></li>
<li><p>Reflect on the data integration process and its importance for decision making</p></li>
<li><p>Recognize activities that highlight the value of data integration for genomic surveillance</p></li>
</ol>
<p><br>
<br>
</p>
</section>
<section id="case-study-sars-cov-2-surveillance-in-romania-during-the-first-months-of-the-pandemic" class="level1">
<h1>Case Study: SARS-CoV-2 surveillance in Romania during the first months of the pandemic</h1>
<p><br>
</p>
<p>Note: This case-study is based on real sequence data from&nbsp; “<strong>Molecular epidemiology analysis of SARS-CoV-2 strains circulating in Romania during the first months of the pandemic</strong>.” by Surleac, Marius, et al., <a href="https://doi.org/10.3390/life10080152" class="uri">https://doi.org/10.3390/life10080152</a> , but it has been modified for the purposes of teaching this course.&nbsp;</p>
<section id="input-files" class="level2">
<h2 class="anchored" data-anchor-id="input-files"><strong>Input files:</strong></h2>
<p><a href="https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa?usp=sharing" class="uri">https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa?usp=sharing</a></p>
</section>
<section id="resources-needed" class="level2">
<h2 class="anchored" data-anchor-id="resources-needed">Resources needed</h2>
<p>Internet access, Projector &amp; screen, Participants should bring their laptops equipped with Google Chrome or Mozilla Firefox.</p>
<p>Access to: <a href="https://microreact.org/" class="uri">https://microreact.org/</a>, <a href="https://pathogen.watch/" class="uri">https://pathogen.watch/</a>, <a href="https://data-flo.io/" class="uri">https://data-flo.io/</a></p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Note: This case-study is based on real sequence data from&nbsp;the article “<strong>Molecular epidemiology analysis of SARS-CoV-2 strains circulating in Romania during the first months of the pandemic</strong>.” by Surleac, Marius, et al., <a href="https://doi.org/10.3390/life10080152" class="uri">https://doi.org/10.3390/life10080152</a> , but it has been modified for the purposes of teaching this course.&nbsp;</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/0 workflow.png" class="img-fluid figure-img"></p>
<figcaption>The workflow of this exercise</figcaption>
</figure>
</div>
<p><strong><em>Explore the epidemiological information related to 24 SARS-CoV-2 samples from Romania.</em></strong></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th><strong>Country</strong></th>
<th><strong>Town</strong></th>
<th><strong>NAME</strong></th>
<th><strong>Date</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468134</td>
<td>21/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468135</td>
<td>22/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468136</td>
<td>22/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468137</td>
<td>22/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468138</td>
<td>25/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468139</td>
<td>25/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468140</td>
<td>25/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468141</td>
<td>25/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468142</td>
<td>25/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468143</td>
<td>25/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468144</td>
<td>25/03/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468145</td>
<td>28/03/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468146</td>
<td>08/04/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468147</td>
<td>11/04/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468148</td>
<td>11/04/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468149</td>
<td>12/04/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468150</td>
<td>12/04/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468151</td>
<td>15/04/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468152</td>
<td>16/04/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468153</td>
<td>17/04/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468154</td>
<td>17/04/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Bucuresti</td>
<td>EPI_ISL_468155</td>
<td>18/04/2020</td>
</tr>
<tr class="odd">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468157</td>
<td>29/05/2020</td>
</tr>
<tr class="even">
<td>Romania</td>
<td>Suceava</td>
<td>EPI_ISL_468158</td>
<td>29/05/2020</td>
</tr>
</tbody>
</table>
<p>The genomes of the isolates presented in the table above have been sequenced, assembled and the assemblies have been uploaded to Pathogenwatch for further analyses.</p>
<p><strong><em>Activity 1. Data analysis of 24 SARS-CoV-2 genomes from Romania, collected during the early months of the pandemic.</em></strong></p>
<p>Explore a collection of 24 SARS-CoV-2 genomes in Pathogenwatch.</p>
<p><a href="https://pathogen.watch/collection/5hpqvjar7gfa-sars-cov-2-romania-v1" class="uri">https://pathogen.watch/collection/5hpqvjar7gfa-sars-cov-2-romania-v1</a></p>
<p>Q1: In the statistics tab, check the genome lengths. Why do you think these have variable values?</p>
<p>Q2: Why are the values of the N50 equal to the values of the genome lengths?</p>
<p>Q3: What do you think of the QC statistics of the genome assemblies overall?</p>
<p>Q4: Go to the Typing tab. What do you think is a “lineage” of SARS-CoV-2? Why were “lineages” defined early in the pandemic? What lineages do these genomes belong to?</p>
<p>Q5: Go to the “Notable mutations” tab. Observe that all the genomes in this collection belong to the same lineage and they all share the same notable mutations. What does this tell you about the diversity of this collection?</p>
<p>Q6: Click on one of the genomes and see the “Genome card”. Check the details of this lineage in a global context, by pressing <strong><em>View B.1.1 in the COG-UK Global Microreact</em></strong>. Where did this lineage spread and when?</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/1. genome card.png" class="img-fluid figure-img" width="375"></p>
<figcaption>Pathogenwatch screenshot showing a genome card for a SARS-CoV-2 genome from the lineage B1.1 . Notice the <strong><em>View B.1.1 in the COG-UK Global Microreact</em></strong> link</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/2. B1.1 global spread.png" class="img-fluid figure-img"></p>
<figcaption>Microreact view showing isolates from the B1.1 lineage.</figcaption>
</figure>
</div>
<p><strong><em>Activity 2. Data integration</em></strong></p>
<p>In this activity, you will merge the epidemiological information related to the isolates, the bioinformatics analysis files produced by Pathogenwatch and with the tree in Newick format, to produce a Microreact.</p>
<p>Download all the files used in this exercise from <a href="https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa" class="uri">https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa</a></p>
<p>Go to Data-flo: <a href="https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm" class="uri">https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm</a></p>
<p>Click on the Run button. Load the input files. Notice that you need an “access token” from Microreact.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/3. data-flo run.png" class="img-fluid figure-img" width="400"></p>
<figcaption>The RUN button in Data-flo</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/4. Microreact access token.png" class="img-fluid figure-img" width="440"></p>
<figcaption>You will find your access token in your Microreact account, under “Account settings”</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/5. data-flo output.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption>The output of data-flo is a link to a microreact</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/6. microreact.png" class="img-fluid figure-img"></p>
<figcaption>The microreact produced by the data-flo</figcaption>
</figure>
</div>
<p>Copy the data-flo to your account.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/7 copy data-flo.png" class="img-fluid figure-img"></p>
<figcaption>Once you are logged in to your data-flo account, you will be able to copy a public data-flo to your own account. Press on the symbol showing two overlapping papers on the right top corner to copy the current data-flo to your account</figcaption>
</figure>
</div>
<p>Run the data-flo in debug mode and explore how data is being transformed between steps.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/8. data-flo debugger.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>Press the bug symbol at the top left corner to trigger the Debug mode. Notice the Data-flo debugger loading at the bottom of the screen.</figcaption>
</figure>
</div>
<p>Q1: Which adaptors are responsible for data entry?</p>
<p>Q2: What does the “Forward geocoding” adaptor do?</p>
<p>Q3: Check the Microreact link produced by the data-flo.</p>
<p><strong><em>Activity 3. Data visualisation with Microreact</em></strong></p>
<p>Explore the Microreact produced by the Data-flo in the previous step.</p>
<p>Colour the leaves of the tree by “Town” and display the metadata columns for “Notable mutations” and the “Pangolin lineage”.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/9. Microreact colour leaves.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>In order to set the colour of the leaves of a tree in Microreact, go to the eye symbol on top right and choose “Town” in the “Colour Column”</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/10 select metadata block.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>Select which columns you would like to display in the metadata blocks</figcaption>
</figure>
</div>
<p>Your personal Microreact should like this one: <a href="https://microreact.org/project/sars-cov-2-romania1" class="uri">https://microreact.org/project/sars-cov-2-romania1</a></p>
<p>Q1: Look at the timelime. Which is the first isolate sampled from this collection? When was it sampled and where?</p>
<p>Q2: When was the first isolate from the Bucharest sampled and when was the first isolate from Suceava sampled?</p>
<p><strong><em>Activity 4. Data analysis with context genomes</em></strong></p>
<p>The bioinformatician in your team has found another 26 SARS-CoV-2 genomes from Romania in a public database and wants to include them in your study.</p>
<p>View the whole collection of 51 genomes in Pathogenwatch <a href="https://pathogen.watch/collection/gqhf7ndghsim-test2" class="uri">https://pathogen.watch/collection/gqhf7ndghsim-test2</a></p>
<p>Select one of the genomes that belong to the lineage B.1.408 and open the “Genome card”.</p>
<p>Then click on the <strong><em>View B.1.408 in the COG-UK Global Microreact</em></strong></p>
<p>Q1: How many entries of this sublineage are found in the database?</p>
<p>Q2: In what countries was this sub-lineage found ? What does this tell you about the transmission links between Romania and those countries?</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/11 lineage B1408.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>Lineage B.1.408 shown in Microreact</figcaption>
</figure>
</div>
<p>Select the genome that belongs to the B.1.520 lineage, open its “Genome card” and click on the <strong><em>View B.1.520 in the COG-UK Global Microreact.</em></strong></p>
<p>Q3: In what countries was this sub-lineage found ? What does this tell you about the transmission links between Romania and those countries?</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/12 lineage B.1.520.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>Lineage B.1.520 shown in Microreact</figcaption>
</figure>
</div>
<p><strong><em>Activity 5. Data integration for the collection of 51 genomes.</em></strong></p>
<p>Load the same Data-flo as in Activity 2: <a href="https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm" class="uri">https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm</a></p>
<p>Run the data-flo with the new set of input files and load the new Microreact.</p>
<p><strong><em>Activity 6. Explore the new Microreact.</em></strong></p>
<p>Colour the nodes of the tree by “Town” and display the mutations in the metadata blocks.</p>
<p>Your microreact should look like this: <a href="https://microreact.org/project/sars-cov-2-romania-2" class="uri">https://microreact.org/project/sars-cov-2-romania-2</a></p>
<p>Q1: According to this study, where did SARS-CoV-2 spread first outside of Suceava and Bucuresti?</p>
<p>Q2: Where was the isolate belonging to the B.1.520 lineage found in Romania?</p>
<p><strong><em>Activity 7. Discuss the Romanian sequences in a global context.</em></strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/13 global.jpeg" class="img-fluid figure-img"></p>
<figcaption>Figure from the original study published at <a href="https://doi.org/10.3390/life10080152" class="uri">https://doi.org/10.3390/life10080152</a></figcaption>
</figure>
</div>
<p>Q1: How many introductions of SARS-CoV-2 into Romania can you identify based on this figure?</p>
<p>Q2: Why do you think that the isolates from Suceava are interleaved with isolates from abroad?</p>
</section>
<section id="appendix" class="level2">
<h2 class="anchored" data-anchor-id="appendix">Appendix</h2>
<p>Data-flo Documentation: <a href="https://docs.data-flo.io/introduction/readme" class="uri">https://docs.data-flo.io/introduction/readme</a></p>
<p>If you found data-flo useful and would like to continue using it, start by creating an account <a href="https://data-flo.io/signin" class="uri">https://data-flo.io/signin</a>. There are a few public workflows with useful transformations such as left join two CSVs (<a href="https://data-flo.io/run?bzYR3DtxRJsBty5ezb5LoJ" class="uri">https://data-flo.io/run?bzYR3DtxRJsBty5ezb5LoJ</a>) or the Geocoder (<a href="https://data-flo.io/run?kvpsi3T8V" class="uri">https://data-flo.io/run?kvpsi3T8V</a>) that you can try.<br>
</p>
<p>A similar workflow to the one used on this course can be found in the documentation <a href="https://docs.data-flo.io/tutorials/prep-outbreak-data-for-microreact" class="uri">https://docs.data-flo.io/tutorials/prep-outbreak-data-for-microreact</a>. In this workflow, one of the inputs is a SNP matrix.&nbsp;</p>
<p>Data-flo also allows you to import data from a Google Spreadsheet (<a href="https://docs.data-flo.io/using-data-flo/specific-adaptors/google-spreadsheet" class="uri">https://docs.data-flo.io/using-data-flo/specific-adaptors/google-spreadsheet</a>) or from a database (<a href="https://docs.data-flo.io/tutorials/common-use-cases-solved/connect-directly-to-a-database" class="uri">https://docs.data-flo.io/tutorials/common-use-cases-solved/connect-directly-to-a-database</a>) (as opposed to downloading data from the spreadsheet/database into a file).&nbsp;<br>
</p>
<p>Microreact documentation: <a href="https://docs.microreact.org/" class="uri">https://docs.microreact.org/</a></p>
<p>To be able to edit and manage projects, please create your own Microreact Account <a href="https://microreact.org/api/auth/signin" class="uri">https://microreact.org/api/auth/signin</a>.&nbsp;</p>
<p>You can customise the colours of your visualisations by assigning colours to the different variables before you create the Microreact project and/or after the project was created <a href="https://docs.microreact.org/instructions/labels-colours-and-shapes" class="uri">https://docs.microreact.org/instructions/labels-colours-and-shapes</a>.</p>
<p>Everyone with access to a Microreact project has access to the data the project uses. Privacy and permissions can be configured to change who can access the project and its data <a href="https://docs.microreact.org/instructions/access-control-and-project-sharing" class="uri">https://docs.microreact.org/instructions/access-control-and-project-sharing</a>.</p>
<p><br>
</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>This work is licensed under a Creative Commons Attribution 4.0 International License. Reuse is encouraged with acknowledgement</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>