<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Module 2 – Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../agenda.html"> 
<span class="menu-text">Agenda</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-training-materials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Training materials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-training-materials">    
        <li>
    <a class="dropdown-item" href="../Day1/QC.html">
 <span class="dropdown-text">QC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day1/Assembly.html">
 <span class="dropdown-text">Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/Annotation.html">
 <span class="dropdown-text">Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/AMR-detection.html">
 <span class="dropdown-text">AMR-detection</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Phylogenetics.html">
 <span class="dropdown-text">Phylogenetics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Genomic_epidemiology.html">
 <span class="dropdown-text">Genomic Epidemiology for Outbreak Investigation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/GenEpi_Surveillance.html">
 <span class="dropdown-text">Genomic Epidemiology for Pathogen Surveillance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Assessment/assessment.html">
 <span class="dropdown-text">Assessment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Module 2</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Module 2</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="module-2-data-retrieval-read-qc-genome-assemblies-and-assemblies-qc" class="level3">
<h3 class="anchored" data-anchor-id="module-2-data-retrieval-read-qc-genome-assemblies-and-assemblies-qc">Module 2: Data retrieval, Read QC, Genome Assemblies and Assemblies QC</h3>
<section id="file-formats" class="level4">
<h4 class="anchored" data-anchor-id="file-formats">File formats</h4>
<p>Read more here:</p>
<p><a href="https://genome.ucsc.edu/FAQ/FAQformat.html" class="uri">https://genome.ucsc.edu/FAQ/FAQformat.html</a></p>
</section>
<section id="connect-to-your-amazon-ec2-instance" class="level4">
<h4 class="anchored" data-anchor-id="connect-to-your-amazon-ec2-instance">Connect to your Amazon EC2 instance</h4>
<p>Follow the steps in Module 1 to connect to your remote virtual machine.</p>
<p>The next exercise is broadly based on the study <a href="https://journals.asm.org/doi/full/10.1128/msphere.00185-23https://journals.asm.org/doi/full/10.1128/msphere.00185-23">https://journals.asm.org/doi/full/10.1128/msphere.00185-23</a> by Abrudan and Shamanna, 2023</p>
</section>
<section id="data-retrieval-from-the-european-nucleotide-archive-ena" class="level4">
<h4 class="anchored" data-anchor-id="data-retrieval-from-the-european-nucleotide-archive-ena">Data retrieval from the European Nucleotide Archive (ENA)</h4>
<p>The European Nucleotide Archive (ENA) provides a comprehensive record of the world’s nucleotide sequencing information, covering raw sequencing Data, sequence assembly information and functional annotation.</p>
<p>Access to ENA Data is provided through the browser, through search tools, through large scale file download and through the API.</p>
</section>
</section>
<section id="download-the-data-programatically" class="level3">
<h3 class="anchored" data-anchor-id="download-the-data-programatically">Download the Data programatically</h3>
<ol type="1">
<li><p>Connect to your Amazon EC2 Ubuntu instance via ssh.</p>
<p>How to retrieve Data for the samples G18255819 and G18252308 from ENA</p></li>
<li><p>Create a directory called Data in /home/ubuntu/</p></li>
</ol>
<p><code>mkdir /home/ubuntu/</code></p>
<ol start="3" type="1">
<li><p>Create subdirectories for each of your samples</p>
<p><code>mkdir /home/ubuntu/Data/G18255819/illumina</code></p>
<p><code>mkdir /home/ubuntu/Data/G18255819/nanopore</code></p>
<p><code>mkdir /home/ubuntu/Data/G18252308/illumina</code></p>
<p><code>mkdir /home/ubuntu/Data/G18252308/nanopore</code><br>
</p></li>
<li><p>&nbsp;Download the reads for each sample</p></li>
</ol>
<p><code>cd /home/ubuntu/Data/G18255819/nanopore</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR818/004/ERR8187234/ERR8187234.fastq.gz</code></p>
<p><code>cd /home/ubuntu/Data/G18255819/illumina</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR478/004/ERR4784794/ERR4784794_1.fastq.gz</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR478/004/ERR4784794/ERR4784794_2.fastq.gz</code><br>
</p>
<p><code>cd /home/ubuntu/Data/G18252308/nanopore</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR818/005/ERR8187245/ERR8187245.fastq.gz</code></p>
<p><code>cd /home/ubuntu/Data/G18252308/illumina</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR463/006/ERR4635696/ERR4635696_1.fastq.gz</code></p>
<p><code>wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR463/006/ERR4635696/ERR4635696_2.fastq.gz</code></p>
</section>
<section id="download-data-through-the-browser" class="level3">
<h3 class="anchored" data-anchor-id="download-data-through-the-browser">Download Data through the browser</h3>
<ol type="1">
<li>Connect to your virtual machine via the browser of your host machine.</li>
<li>In your virtual machine, open a browser.</li>
<li>Navigate to <a href="https://www.ebi.ac.uk/ena/browser/home" class="uri">https://www.ebi.ac.uk/ena/browser/home</a></li>
<li>Download files associated with the samples G18255819 and G18252308. Downaload the Nanopore and the Illumina reads.</li>
</ol>
</section>
<section id="reads-qc" class="level2">
<h2 class="anchored" data-anchor-id="reads-qc">Reads QC</h2>
<p>In this part of the exercise, we will use a programme called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.</p>
<p>FastQC aims to provide a simple way to do some quality control checks on raw sequence Data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your Data has any problems of which you should be aware before doing any further analysis.</p>
<p>The main functions of FastQC are</p>
<ul>
<li><p>Import of Data from BAM, SAM or FastQ files (any variant)</p></li>
<li><p>Providing a quick overview to tell you in which areas there may be problems</p></li>
<li><p>Summary graphs and tables to quickly assess your Data</p></li>
<li><p>Export of results to an HTML based permanent report</p></li>
</ul>
<section id="run-fastqc-programatically" class="level3">
<h3 class="anchored" data-anchor-id="run-fastqc-programatically">Run FastQC programatically</h3>
<p>To run non-interactively you simply have to specify a list of files to process on the command line.&nbsp;</p>
<p><code>fastqc somefile.txt someotherfile.txt</code></p>
<p><code>fastqc /home/ubuntu/Data/*/*/*</code></p>
</section>
<section id="qc-your-fastq-reads-through-the-visual-interface" class="level3">
<h3 class="anchored" data-anchor-id="qc-your-fastq-reads-through-the-visual-interface">QC your fastq reads through the visual interface</h3>
<p>Navigate to <em>/home/ubuntu/Software/FastQC</em> and double click on the FastQC icon.</p>
<p>In the visual interface, open all files produced by FastQC and assess the results, eg ERR4635696_1_fastqc.html</p>
<p>Visualize and interpret the FastQC results. Compare the QC results for Nanopore and Illumina sequence reads</p>
<p><strong>Optional</strong>: Unzip the files <em>/home/ubuntu/Data/*/*/*_*_fastqc.zip</em> and produce a script that extracts the information on reads lengths, GC contents etc; plot the results from all samples.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/fastqc_illumina.png" class="img-fluid figure-img" width="607"></p>
<figcaption>FastQC Illumina reads</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/fastqc_nanopore.png" class="img-fluid figure-img" width="560"></p>
<figcaption>FastQC Nanopore reads</figcaption>
</figure>
</div>
</section>
</section>
<section id="speciation" class="level2">
<h2 class="anchored" data-anchor-id="speciation">Speciation</h2>
<p><a href="https://gitlab.com/antunderwood/bactinspector">Bactinspector</a> is a package to a) determine the most probable species based on sequence in fasta/fastq files using refseq and Mash (<a href="https://mash.readthedocs.io/en/latest/index.html" class="uri">https://mash.readthedocs.io/en/latest/index.html</a>) and b) determine the closest reference in refseq to a set of fasta/fastq files.</p>
<ol type="1">
<li>Go to sample files /<em>home/ubuntu/Data/</em> and run Bactinspector using the following command</li>
</ol>
<p><code>bactinspector closest_match -fq "*_2.fastq.gz"</code></p>
<p><img src="images/bactinspector.png" class="img-fluid"></p>
<ol start="2" type="1">
<li>Tidy up the output *.tsv file</li>
</ol>
<p><code>tail -n +2 *.tsv| column -t | less -S</code><br>
<code>cat *.tsv | tr "\t" "~" | cut -d"~" -f2</code></p>
<p><em>Eg of a result:</em></p>
<p><em>ASM289538v1</em></p>
<p><em>GCF_002895385</em><br>
</p>
<ol start="3" type="1">
<li><p>How do you interpret the results? Look up your results in ENA.</p>
<p><em>In your browser, type</em></p>
<p><a href="https://www.ebi.ac.uk/ena/browser/text-search?query=ASM289538v1"><em>https://www.ebi.ac.uk/ena/browser/text-search?query=ASM289538v1</em></a></p>
<p><img src="images/ENA search result.png" class="img-fluid"></p></li>
</ol>
<p>Check fields such as:</p>
<ul>
<li><p><strong>Assembly Level</strong></p></li>
<li><p><strong>Genome Representation</strong></p></li>
<li><p><strong>Strain</strong></p></li>
<li><p><strong>Count Contig</strong></p></li>
<li><p><strong>Contig N50</strong></p></li>
<li><p><strong>Total Length</strong></p></li>
</ul>
<p>Download the reference genome from the visual reference</p>
<p><img src="images/download reference.png" class="img-fluid" width="236"></p>
<p>Or from the command line.</p>
<p><code>wget https://www.ebi.ac.uk/ena/browser/api/fasta/CP018629.1 -O - &gt;&gt; GCA_002895385-chromosomes.fasta</code></p>
</section>
<section id="genome-assembly" class="level2">
<h2 class="anchored" data-anchor-id="genome-assembly">Genome assembly</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/N50.jpg" class="img-fluid figure-img"></p>
<figcaption>An example of the N50 metric</figcaption>
</figure>
</div>
<p><strong>The Velvet assembler</strong></p>
<p><strong>Velvet</strong> is an algorithm package that has been designed to deal with <em>de novo</em> <a href="https://en.wikipedia.org/wiki/Genome_assembly" title="Genome assembly">genome assembly</a> and short read <a href="https://en.wikipedia.org/wiki/Sequence_alignment" title="Sequence alignment">sequencing alignments</a>. This is achieved through the manipulation of <a href="https://en.wikipedia.org/wiki/De_Bruijn_graph" title="De Bruijn graph">de Bruijn graphs</a> for genomic sequence assembly via the removal of errors and the simplification of repeated regions.</p>
<p><strong>Run Velvet</strong></p>
<p>Find a value of <em>k</em> (between 21 and 99) to start with, and record your choice.</p>
<p><code>velveth G18252308 KMER -short -separate -fastq /home/ubuntu/Data/G18252308/illumina/ERR4635696_1.fastq.gz /home/ubuntu/Data/G18252308/illumina/ERR4635696_2.fastq.gz</code></p>
<p>After <code>velveth</code> is finished, look in the new folder that has the name you chose. You should see the following files:</p>
<pre><code>Log
Roadmaps
Sequences</code></pre>
<p>The ’<code>Log</code>‘ file has a useful reminder of what commands you typed to get this assembly result, for reproducing results later on. ’<code>Sequences</code>‘ contains the sequences we put in, and ’<code>Roadmaps</code>‘ contains the index you just created.</p>
<p>Now we will run the assembly with default parameters:</p>
<pre><code>velvetg G18252308</code></pre>
<p><img src="images/velveth.png" class="img-fluid"></p>
<p>Velvet will end with a text like this:</p>
<p><code>Final graph has ... nodes and n50 of ..., max ..., total ..., using .../... reads</code></p>
<p><img src="images/velvetg.png" class="img-fluid"></p>
<p>The number of nodes represents the number of nodes in the graph, which (more or less) is the number of contigs. Velvet reports its N50 (as well as everything else) in ‘kmer’ space. The conversion to ‘basespace’ is as simple as adding k-1 to the reported length.</p>
<p>Look again at the folder <code>asm_name</code>, you should see the following extra files:</p>
<p><code>contigs.faGraphLastGraphPreGraphstats.txt</code></p>
<p>The important files are:</p>
<p><code>contigs.fa</code> - the assembly itself<code>Graph</code> - a textual representation of the contig graph<code>stats.txt</code> - a file containing statistics on each contig</p>
<p><strong>Questions</strong></p>
<ul>
<li><p>What k-mer did you use?</p></li>
<li><p>What is the N50 of the assembly?</p></li>
<li><p>What is the size of the largest contig?</p></li>
<li><p>How many contigs are there in the <code>contigs.fa</code> file? Use <code>grep -c NODE contigs.fa</code></p></li>
</ul>
<p>Run again <strong>velveth</strong> and <strong>velvetg</strong> with another KMER size.</p>
<p><img src="images/velveth-kmer51.png" class="img-fluid"></p>
<p><strong><em>In the classroom, decide you got the best N50 and the lowest number of contigs!</em></strong></p>
<p><strong>Visualise the assembles using ACT and Artemis</strong></p>
<p>ACT is a Java application for displaying pairwise comparisons between two or more DNA sequences.</p>
<p>ACT can be used to identify and analyse regions of similarity and difference between genomes and to explore conservation of synteny, in the context of the entire sequences and their annotation. It can read complete EMBL, GENBANK and GFF entries or sequences in FASTA or raw format.</p>
<p><em>Tip: ACT and Artemis are installed here: /usr/share/miniconda/pkgs/artemis-18.2.0-hdfd78af_0/share/artemis-18.2.0-0</em>&nbsp;</p>
<p>Go the the visual interface of your virtual machine.</p>
<p>In the Terminal, type</p>
<p><code>art &amp;</code></p>
<p>Artemis should open. Load your contigs.fa file.</p>
<p><img src="images/artemis interface.png" class="img-fluid" width="346"></p>
<p><img src="images/artemis assembly.png" class="img-fluid" width="631"></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/artemis overview.png" class="img-fluid figure-img" width="304"></p>
<figcaption>Try View-&gt;Overview in the Artemis menu</figcaption>
</figure>
</div>
<p>From the <em>Graph</em> menu, open <em>GC Deviation (G-C)/(G+C)</em> by clicking on the button next to it.</p>
<p>Rescale the plot for to a more appropriate window size for this zoomed out view: Right click on the graph, and click <em>Maximum Window Size</em>, and select <em>20000</em>. Then move the graph slider of the right hand side of the screen down to the bottom of the bar.</p>
<p><img src="images/artemis gc.png" class="img-fluid"></p>
<p>From the graph you can see that plot generally varies about an upper level and a lower level across the assembly, with shifts occurring at contig boundaries.</p>
</section>
<section id="assembly-qc-with-quast" class="level2">
<h2 class="anchored" data-anchor-id="assembly-qc-with-quast">Assembly QC with Quast</h2>
<p><a href="https://github.com/ablab/quast">QUAST</a> stands for QUality ASsessment Tool. It evaluates genome/metagenome assemblies by computing various metrics. The current QUAST toolkit includes the general QUAST tool for genome assemblies, MetaQUAST, the extension for metagenomic Datasets, QUAST-LG, the extension for large genomes (e.g., mammalians), and Icarus, the interactive visualizer for these tools.</p>
<p><img src="images/quast.png" class="img-fluid"></p>
<p>In your virtual machine, go to the Chrome browser, and navigate to:</p>
<p><a href="file:///home/ubuntu/Data/G18252308/illumina/G18252308_quast/report.html"><em>file:///home/ubuntu/Data/G18252308/illumina/G18252308_quast/report.html</em></a></p>
<p>Alternatively, navigate to <em>/home/ubuntu/Data/G18252308/illumina/G18252308_quast</em> and double click on <em>report.html</em></p>
<p><img src="images/quast viz.png" class="img-fluid"></p>
<p>What is the cumulative length of the assembly? Can you check if this expected for a <em>S. aureus</em> genome? <strong>Hint</strong>: go back to the reference genome you found previously! What was the length of that assembly?</p>
</section>
<section id="assembly-methods-comparison" class="level2">
<h2 class="anchored" data-anchor-id="assembly-methods-comparison">Assembly methods comparison</h2>
<p>For assembly of short (Illumina) reads, we will use <code>Unicycler</code>. With short read data, <code>Unicycler</code> acts as wrapper script for the <code>SPAdes</code> assembler, and will produce an assembly with settings optimised for bacteria.</p>
<p>In the command below we:</p>
<ul>
<li><p>Allocate 4 CPUs to the assembler</p>
<ul>
<li><code>-t 4</code></li>
</ul></li>
<li><p>Specify the Illumina forward and reverse reads to use</p>
<ul>
<li><code>-1 &lt;fastq&gt;</code> and <code>-2 &lt;fastq&gt;</code></li>
</ul></li>
<li><p>Specify a directory to output files to</p>
<ul>
<li><code>-o &lt;dir&gt;</code></li>
</ul></li>
</ul>
<p><code>unicycler -t 4 -1 /home/ubuntu/Data/G18252308/illumina/ERR4635696_1.fastq.gz -2 /home/ubuntu/Data/G18252308/illumina/ERR4635696_2.fastq.gz -o ERR4635696.short</code></p>
<p>For assembly of long reads, <code>Unicycler</code> switches to the long read assembler <code>miniasm</code> - this is a very rapid assembler, but is not particularly accurate, and can often introduce errors and mistakes into its outputs.<br>
</p>
<p>In the command below we:</p>
<ul>
<li><p>Allocate 4 CPUs to the assembler</p>
<ul>
<li><code>-t 4</code></li>
</ul></li>
<li><p>Specify the ONT long reads to use</p>
<ul>
<li><code>-l &lt;fastq&gt;</code></li>
</ul></li>
<li><p>Specify a directory to output files to</p>
<ul>
<li><code>-o &lt;dir&gt;</code></li>
</ul></li>
</ul>
<p><code>unicycler -t 4 -l /home/ubuntu/Data/G18252308/nanopore/ERR8187245.fastq.gz -o S_aureus_JKD6159.s100.unicycler.long</code></p>
<p>For assembly of both long and short reads, <code>Unicycler</code> uses <code>SPAdes</code> for an initial ‘short read’ assembly, and then uses the long reads to attempt to bridge the gaps between contigs from the initial <code>SPAdes</code> assembly. This can work well when the short read data is of high quality/depth but the ONT data is of low depth. However, the ‘bridging step’ can be time consuming, and if the short read assembly is poor, the final assembly may also be. If you have high quality and depth long reads, this may not be the best option.</p>
<p>In the command below we:</p>
<ul>
<li><p>Allocate 4 CPUs to the assembler</p>
<ul>
<li><code>-t 4</code></li>
</ul></li>
<li><p>Specify the Illumina forward and reverse reads to use</p>
<ul>
<li><code>-1 &lt;fastq&gt;</code> and <code>-2 &lt;fastq&gt;</code></li>
</ul></li>
<li><p>Specify the ONT long reads to use</p>
<ul>
<li><code>-l &lt;fastq&gt;</code></li>
</ul></li>
<li><p>Specify a directory to output files to</p>
<ul>
<li><code>-o &lt;dir&gt;</code></li>
</ul></li>
</ul>
<p>Note, this method may take over an hour to run on the VMs - check with an instructor before running.</p>
<p><code>unicycler -t 4 -1/home/ubuntu/Data/G18252308/illumina/ERR4635696_1.fastq.gz -2 /home/ubuntu/Data/G18252308/illumina/ERR4635696_2.fastq.gz -l /home/ubuntu/Data/G18252308/nanopore/ERR8187245.fastq.gz -o ERR8187245.unicycler.hybrid</code></p>
</section>
<section id="compare-velvet-and-spades-assemblers" class="level2">
<h2 class="anchored" data-anchor-id="compare-velvet-and-spades-assemblers">Compare Velvet and SPADES assemblers</h2>
<p>Run SPADES (with Unicycler)</p>
<p><code>unicycler -t 4 -1 /home/ubuntu/Data/G18255819/illumina/ERR4784794_1.fastq.gz -2 /home/ubuntu/Data/G18255819/illumina/ERR4784794_2.fastq.gz -o ERR4784794.short</code></p>
<p>View the SPADES assembly metrics in Quast</p>
<p><img src="images/quast _ spades.png" class="img-fluid" width="734"></p>
<p>How do these metrics compare with the results you got from the Velvet assembler?</p>
</section>
<section id="genome-annotation-with-prokka" class="level2">
<h2 class="anchored" data-anchor-id="genome-annotation-with-prokka">Genome annotation with Prokka</h2>
<section id="run-prokka-for-one-sample" class="level4">
<h4 class="anchored" data-anchor-id="run-prokka-for-one-sample">Run Prokka for one sample</h4>
<p><code>cd /home/ubuntu/Data/</code></p>
<p><code>docker run -v $(pwd):$(pwd) -w $(pwd) staphb/prokka:latest prokka --outdir /home/ubuntu/Data/annotations/G18252308 --prefix G18252308 /home/ubuntu/Data/G18252308/illumina/G18252308_kmer51/contigs.fa</code></p>
<p>Check if the <em>mecA</em> gene was found in the annotated genome.</p>
<p><img src="images/grep mecA.png" class="img-fluid"></p>
<p>Once prokka is done, open Artemis and load the .gff file and locate the <em>mecA</em> gene. How many nucleotides does it have?</p>
<p><img src="images/artemis mecA.png" class="img-fluid"></p>
</section>
<section id="how-to-run-prokka-on-all-samples-from-one-folder" class="level4">
<h4 class="anchored" data-anchor-id="how-to-run-prokka-on-all-samples-from-one-folder">How to run Prokka on all samples from one folder</h4>
<p>Example script: create the following bash file: <code>/home/ubuntu/Data/PROKKA_run/run_prokka.sh</code></p>
<p>Add the following lines of code:</p>
<p>for sample in `ls /home/ubuntu/Data/all_fastas/*.fasta | sed ‘s/\.fasta//’`</p>
<p>do</p>
<p>&nbsp;echo $sample</p>
<p>&nbsp;output=$(echo $sample | sed -E ‘s#.*/([^/]+)$#\1#’)</p>
<p>&nbsp;echo $output</p>
<p>&nbsp;docker run -v $(pwd):$(pwd) -w $(pwd) staphb/prokka:latest prokka –outdir /home/ubuntu/Data/annotations/${output} –prefix ${output} /home/ubuntu/Data/all_fastas/${output}.fasta</p>
<p>done</p>
</section>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"></h2>
</section>
<section id="bonus-activity" class="level2">
<h2 class="anchored" data-anchor-id="bonus-activity">Bonus activity</h2>
<p>Install <a href="https://github.com/tseemann/snippy">Snippy</a>, snp_sites and IQtree and run a SNP tree. See the tutorial <a href="https://github.com/WCSCourses/GenEpiLAC2024/blob/main/Manuals/Mapping_and_Phylogenetics/Mapping%2BPhylo.md">here</a></p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>This work is licensed under a Creative Commons Attribution 4.0 International License. Reuse is encouraged with acknowledgement</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>