---
title: "Module 4"
---

SARS-CoV-2 surveillance in Romania

-   This exercise used tools developed by The Centre for Genomic Pathogen Surveillance (CGPS) at University of Oxford

-   Similar exercises have been developed as part of T3Connect <https://wcscourses.github.io/T3connectResources/>

Genomic surveillance of pathogens aims to understand the emergence and dissemination of pathogens or their lineages of risk with the ultimate goal of implementing evidence-based interventions to protect public health. Epidemiological data from patients is collected by healthcare professionals. Data on species identification and any phenotypic or molecular characterization of the isolates is often generated by the microbiology laboratories linked to healthcare facilities and/or the reference laboratory. In an ideal scenario, the different sources of laboratory data are stored in a centralised surveillance system and database (such as [WHONET](https://whonet.org/)). However, these systems rarely also incorporate genomic data produced by bioinformaticians. 

The Centre for Genomic Pathogen Surveillance develops free web applications for integration and visualisation of surveillance (and other) data, called [data-flo](https://data-flo.io/) and [Microreact](http://www.microreact.org). The aim of this module is to highlight the role of the various analytics and sources of data in pathogen surveillance, and how streamlined data integration is essential for real-time decision making, while also introducing resources that can facilitate the inclusion of this topic in teaching curricula.

\

# Learning Outcomes

At the end of this module participants will be able to:

1.  Distinguish the different components of genomic surveillance and their sources

2.  Identify analytic tools to extract from genomes useful information for surveillance 

3.  Utilise data-flo to join different sources of data

4.  Utilise Microreact to recognize the interplay between different types of data

5.  Reflect on the data integration process and its importance for decision making

6.  Recognize activities that highlight the value of data integration for genomic surveillance

\
\

# Case Study: SARS-CoV-2 surveillance in Romania during the first months of the pandemic

\

Note: This case-study is based on real sequence data from  "**Molecular epidemiology analysis of SARS-CoV-2 strains circulating in Romania during the first months of the pandemic**." by Surleac, Marius, et al., <https://doi.org/10.3390/life10080152> , but it has been modified for the purposes of teaching this course. 

## **Input files:**

<https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa?usp=sharing>

## Resources needed

Internet access, Projector & screen, Participants should bring their laptops equipped with Google Chrome or Mozilla Firefox.

Access to: <https://microreact.org/>, <https://pathogen.watch/>, <https://data-flo.io/>

## Introduction

Note: This case-study is based on real sequence data from the article "**Molecular epidemiology analysis of SARS-CoV-2 strains circulating in Romania during the first months of the pandemic**." by Surleac, Marius, et al., <https://doi.org/10.3390/life10080152> , but it has been modified for the purposes of teaching this course. 

![The workflow of this exercise](images/0%20workflow.png)

***Explore the epidemiological information related to 24 SARS-CoV-2 samples from Romania.***

| **Country** | **Town**  | **NAME**       | **Date**   |
|-------------|-----------|----------------|------------|
| Romania     | Bucuresti | EPI_ISL_468134 | 21/03/2020 |
| Romania     | Bucuresti | EPI_ISL_468135 | 22/03/2020 |
| Romania     | Bucuresti | EPI_ISL_468136 | 22/03/2020 |
| Romania     | Suceava   | EPI_ISL_468137 | 22/03/2020 |
| Romania     | Suceava   | EPI_ISL_468138 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468139 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468140 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468141 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468142 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468143 | 25/03/2020 |
| Romania     | Suceava   | EPI_ISL_468144 | 25/03/2020 |
| Romania     | Bucuresti | EPI_ISL_468145 | 28/03/2020 |
| Romania     | Suceava   | EPI_ISL_468146 | 08/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468147 | 11/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468148 | 11/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468149 | 12/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468150 | 12/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468151 | 15/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468152 | 16/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468153 | 17/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468154 | 17/04/2020 |
| Romania     | Bucuresti | EPI_ISL_468155 | 18/04/2020 |
| Romania     | Suceava   | EPI_ISL_468157 | 29/05/2020 |
| Romania     | Suceava   | EPI_ISL_468158 | 29/05/2020 |

The genomes of the isolates presented in the table above have been sequenced, assembled and the assemblies have been uploaded to Pathogenwatch for further analyses.

***Activity 1. Data analysis of 24 SARS-CoV-2 genomes from Romania, collected during the early months of the pandemic.***

Explore a collection of 24 SARS-CoV-2 genomes in Pathogenwatch.

<https://pathogen.watch/collection/5hpqvjar7gfa-sars-cov-2-romania-v1>

Q1: In the statistics tab, check the genome lengths. Why do you think these have variable values?

Q2: Why are the values of the N50 equal to the values of the genome lengths?

Q3: What do you think of the QC statistics of the genome assemblies overall?

Q4: Go to the Typing tab. What do you think is a "lineage" of SARS-CoV-2? Why were "lineages" defined early in the pandemic? What lineages do these genomes belong to?

Q5: Go to the "Notable mutations" tab. Observe that all the genomes in this collection belong to the same lineage and they all share the same notable mutations. What does this tell you about the diversity of this collection?

Q6: Click on one of the genomes and see the "Genome card". Check the details of this lineage in a global context, by pressing ***View B.1.1 in the COG-UK Global Microreact***. Where did this lineage spread and when?

![Pathogenwatch screenshot showing a genome card for a SARS-CoV-2 genome from the lineage B1.1 . Notice the ***View B.1.1 in the COG-UK Global Microreact*** link](images/1.%20genome%20card.png){width="375"}

![Microreact view showing isolates from the B1.1 lineage.](images/2.%20B1.1%20global%20spread.png)

***Activity 2. Data integration***

In this activity, you will merge the epidemiological information related to the isolates, the bioinformatics analysis files produced by Pathogenwatch and with the tree in Newick format, to produce a Microreact.

Download all the files used in this exercise from <https://drive.google.com/drive/folders/1yP-ectAQ0plj8G1rcd_UNqPOyFIC7axa>

Go to Data-flo: <https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm>

Click on the Run button. Load the input files. Notice that you need an "access token" from Microreact.

![The RUN button in Data-flo](images/3.%20data-flo%20run.png){width="400"}

![You will find your access token in your Microreact account, under "Account settings"](images/4.%20Microreact%20access%20token.png){width="440"}

![The output of data-flo is a link to a microreact](images/5.%20data-flo%20output.png){width="60%"}

![The microreact produced by the data-flo](images/6.%20microreact.png)

Copy the data-flo to your account.

![Once you are logged in to your data-flo account, you will be able to copy a public data-flo to your own account. Press on the symbol showing two overlapping papers on the right top corner to copy the current data-flo to your account](images/7%20copy%20data-flo.png)

Run the data-flo in debug mode and explore how data is being transformed between steps.

![Press the bug symbol at the top left corner to trigger the Debug mode. Notice the Data-flo debugger loading at the bottom of the screen.](images/8.%20data-flo%20debugger.png){width="40%"}

Q1: Which adaptors are responsible for data entry?

Q2: What does the "Forward geocoding" adaptor do?

Q3: Check the Microreact link produced by the data-flo.

***Activity 3. Data visualisation with Microreact***

Explore the Microreact produced by the Data-flo in the previous step.

Colour the leaves of the tree by "Town" and display the metadata columns for "Notable mutations" and the "Pangolin lineage".

![In order to set the colour of the leaves of a tree in Microreact, go to the eye symbol on top right and choose "Town" in the "Colour Column"](images/9.%20Microreact%20colour%20leaves.png){width="40%"}

![Select which columns you would like to display in the metadata blocks](images/10%20select%20metadata%20block.png){width="40%"}

Your personal Microreact should like this one: <https://microreact.org/project/sars-cov-2-romania1>

Q1: Look at the timelime. Which is the first isolate sampled from this collection? When was it sampled and where?

Q2: When was the first isolate from the Bucharest sampled and when was the first isolate from Suceava sampled?

***Activity 4. Data analysis with context genomes***

The bioinformatician in your team has found another 26 SARS-CoV-2 genomes from Romania in a public database and wants to include them in your study.

View the whole collection of 51 genomes in Pathogenwatch <https://pathogen.watch/collection/gqhf7ndghsim-test2>

Select one of the genomes that belong to the lineage B.1.408 and open the "Genome card".

Then click on the ***View B.1.408 in the COG-UK Global Microreact***

Q1: How many entries of this sublineage are found in the database?

Q2: In what countries was this sub-lineage found ? What does this tell you about the transmission links between Romania and those countries?

![Lineage B.1.408 shown in Microreact](images/11%20lineage%20B1408.png){width="40%"}

Select the genome that belongs to the B.1.520 lineage, open its "Genome card" and click on the ***View B.1.520 in the COG-UK Global Microreact.***

Q3: In what countries was this sub-lineage found ? What does this tell you about the transmission links between Romania and those countries?

![Lineage B.1.520 shown in Microreact](images/12%20lineage%20B.1.520.png){width="40%"}

***Activity 5. Data integration for the collection of 51 genomes.***

Load the same Data-flo as in Activity 2: <https://data-flo.io/run?4LxncBtNr69TmfzTGj9Szm>

Run the data-flo with the new set of input files and load the new Microreact.

***Activity 6. Explore the new Microreact.***

Colour the nodes of the tree by "Town" and display the mutations in the metadata blocks.

Your microreact should look like this: <https://microreact.org/project/sars-cov-2-romania-2>

Q1: According to this study, where did SARS-CoV-2 spread first outside of Suceava and Bucuresti?

Q2: Where was the isolate belonging to the B.1.520 lineage found in Romania?

***Activity 7. Discuss the Romanian sequences in a global context.***

![Figure from the original study published at <https://doi.org/10.3390/life10080152>](images/13%20global.jpeg)

Q1: How many introductions of SARS-CoV-2 into Romania can you identify based on this figure?

Q2: Why do you think that the isolates from Suceava are interleaved with isolates from abroad?

## Appendix

Data-flo Documentation: <https://docs.data-flo.io/introduction/readme>

If you found data-flo useful and would like to continue using it, start by creating an account <https://data-flo.io/signin>. There are a few public workflows with useful transformations such as left join two CSVs (<https://data-flo.io/run?bzYR3DtxRJsBty5ezb5LoJ>) or the Geocoder (<https://data-flo.io/run?kvpsi3T8V>) that you can try.\

A similar workflow to the one used on this course can be found in the documentation <https://docs.data-flo.io/tutorials/prep-outbreak-data-for-microreact>. In this workflow, one of the inputs is a SNP matrix. 

Data-flo also allows you to import data from a Google Spreadsheet (<https://docs.data-flo.io/using-data-flo/specific-adaptors/google-spreadsheet>) or from a database (<https://docs.data-flo.io/tutorials/common-use-cases-solved/connect-directly-to-a-database>) (as opposed to downloading data from the spreadsheet/database into a file). \

Microreact documentation: <https://docs.microreact.org/>

To be able to edit and manage projects, please create your own Microreact Account <https://microreact.org/api/auth/signin>. 

You can customise the colours of your visualisations by assigning colours to the different variables before you create the Microreact project and/or after the project was created <https://docs.microreact.org/instructions/labels-colours-and-shapes>.

Everyone with access to a Microreact project has access to the data the project uses. Privacy and permissions can be configured to change who can access the project and its data <https://docs.microreact.org/instructions/access-control-and-project-sharing>.

\
