<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>mappingphylo – Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Bioinformatic methods for bacterial genomics, Cluj-Napoca, October 2025</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../agenda.html"> 
<span class="menu-text">Agenda</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-training-materials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Training materials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-training-materials">    
        <li>
    <a class="dropdown-item" href="../Day1/QC.html">
 <span class="dropdown-text">QC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day1/Assembly.html">
 <span class="dropdown-text">Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/Annotation.html">
 <span class="dropdown-text">Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day2/AMR-detection.html">
 <span class="dropdown-text">AMR-detection</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Phylogenetics.html">
 <span class="dropdown-text">Phylogenetics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/Genomic_epidemiology.html">
 <span class="dropdown-text">Genomic Epidemiology for Outbreak Investigation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Day3/GenEpi_Surveillance.html">
 <span class="dropdown-text">Genomic Epidemiology for Pathogen Surveillance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Assessment/assessment.html">
 <span class="dropdown-text">Assessment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">




<p><img src="https://coursesandconferences.wellcomeconnectingscience.org/wp-content/themes/wcc_courses_and_conferences/dist/assets/svg/logo.svg" width="200" height="200"></p>
<p><a href="https://github.com/WCSCourses/GenEpiLAC2025/blob/main/course_modules_2025/Manual_main.md">&lt;&lt;&lt; Go back to Manual Contents Page</a></p>
<p><br></p>
<section id="mapping-and-phylogenetics---costa-rica-2025" class="level1">
<h1>Mapping and Phylogenetics - Costa Rica 2025 <!-- omit in toc --></h1>
<section id="module-lead-mat-beale" class="level3">
<h3 class="anchored" data-anchor-id="module-lead-mat-beale">Module Lead: Mat Beale <!-- omit in toc --></h3>
<p><br></p>
</section>
</section>
<section id="table-of-contents" class="level1">
<h1>Table of contents <!-- omit in toc --></h1>
<ul>
<li><a href="#module-overview-and-aims">Module Overview and Aims</a></li>
<li><a href="#introduction">Introduction</a>
<ul>
<li><a href="#mapping-to-reference-genome">Mapping to reference genome</a></li>
<li><a href="#sequencing-data-formats">Sequencing data formats</a></li>
<li><a href="#fastq-format">Fastq format</a></li>
<li><a href="#map-to-reference---sam-format">Map to reference - SAM format</a></li>
<li><a href="#variant-calling">Variant Calling</a></li>
<li><a href="#consensus-calling">Consensus calling</a></li>
</ul></li>
<li><a href="#practical-exercise---mapping--phylogeny">Practical Exercise - Mapping &amp; Phylogeny</a>
<ul>
<li><a href="#finding-the-data">Finding the data</a></li>
<li><a href="#perform-mapping-of-reads-to-reference-genome-using-snippy">Perform mapping of reads to reference genome using snippy</a>
<ul>
<li><a href="#examine-snippy-logs">Examine <code>snippy</code> logs</a></li>
<li><a href="#examine-snippy-outputs">Examine <code>snippy</code> outputs</a></li>
</ul></li>
<li><a href="#map-a-second-genome-to-reference">Map a second genome to reference</a></li>
<li><a href="#Use-an-existing-assembly-as-input-to-snippy">Map an assembly to reference</a>
<ul>
<li><a href="#Map-an-existing-short-read-assembly">Map an existing short-read assembly</a></li>
<li><a href="#Map-a-long-read-assembly-to-reference">Map an existing long-read assembly</a></li>
</ul></li>
<li><a href="#contextualise-new-genomes-with-a-collection">Contextualise new genomes with a collection</a></li>
<li><a href="#phylogenetics">Phylogenetics</a>
<ul>
<li><a href="#make-a-snp-only-alignment-using-snp-sites">Make a SNP-only alignment using <code>snp-sites</code></a></li>
<li><a href="#calculate-a-phylogenetic-tree-from-the-snps-using-iq-tree2">Calculate a phylogenetic tree from the SNPs using <code>IQ-TREE2</code></a></li>
<li><a href="#accounting-for-recombination-with-gubbins">Accounting for recombination with <code>gubbins</code></a></li>
<li><a href="#root-a-phylogeny">Root a phylogeny</a></li>
<li><a href="#clustering-genomes-using-fastbaps">Clustering genomes using <code>fastBAPS</code></a></li>
</ul></li>
</ul></li>
</ul>
<p><br></p>
</section>
<section id="module-overview-and-aims" class="level1">
<h1>Module Overview and Aims</h1>
<p><br></p>
<p>When we sequence a genome, we aim to capture information about: * Single Nucleotide Polymorphisms (SNPs) * Insertions and Deletions (INDELs) * Copy Number Variations (CNVs) * Structural rearrangements * Gene gain and loss (i.e.&nbsp;the pangenome)</p>
<p>There are two main approaches used for reconstructing bacterial genomes for population analyses.</p>
<ul>
<li><em>De novo</em> assembly</li>
<li>Map to Reference genome</li>
</ul>
<p>Each has benefits and limitations. This module will focus on Reference mapping.</p>
<p>For reference mapping, whether we are dealing with different bacterial isolates, with viral populations in a patient, or even with genomes of different human individuals, the principles are essentially the same. Instead of assembling the newly generated sequence reads <em>de novo</em> to produce a new genome sequence, it is easier and much faster to align or map the new sequence data to the reference genome (please note that we will use the terms “aligning” and “mapping” interchangeably). We can then readily identify SNPs and INDELs that distinguish closely related populations or individual organisms and may thus learn about genetic differences that may cause drug resistance or increased virulence in pathogens, or changed susceptibility to disease in humans. One important prerequisite for the mapping of sequence data to work is that the reference and the re-sequenced subject have the same genome architecture.</p>
<p>In this exercise, we will use sequence data from <em>Vibrio cholerae</em> genomes to learn about mapping and phylogenetics. Importantly, although the genome data is real, the collection has been assembled for the purpose of this practical exercise, and has also been edited to make it run more efficiently in the Virtual Machine training environment.</p>
<p><br></p>
</section>
<section id="introduction" class="level1">
<h1><a href="#introduction">Introduction</a></h1>
<section id="mapping-to-reference-genome" class="level2">
<h2 class="anchored" data-anchor-id="mapping-to-reference-genome"><a href="#mapping-to-reference-genome">Mapping to reference genome</a></h2>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Mapping_Overview_diag_2.png" class="img-fluid figure-img"></p>
<figcaption>This is an image</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="sequencing-data-formats" class="level2">
<h2 class="anchored" data-anchor-id="sequencing-data-formats"><a href="#Sequencing-data-formats">Sequencing data formats</a></h2>
<p>Raw data directly from the sequencer may be in sequencer-specific formats, but will always be converted to <code>fastq</code> format for downstream analysis. <code>fastq</code> is a plain text format that is standard within bioinformatics.</p>
<p><br></p>
</section>
<section id="fastq-format" class="level2">
<h2 class="anchored" data-anchor-id="fastq-format">Fastq format</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastq-expanded2.png" class="img-fluid figure-img"></p>
<figcaption>fastq multiline</figcaption>
</figure>
</div>
<p><br></p>
<p>In a fastq file, each sequencing read is represented by 4 different rows</p>
<ul>
<li>Read ID
<ul>
<li>always begins with ‘@’</li>
<li>Often includes information about sequencer and location on flow cell</li>
</ul></li>
<li>Sequence info
<ul>
<li>The DNA sequence (A, T, G, C, N) at each position in read</li>
</ul></li>
<li>Comment Line
<ul>
<li>Begins ‘+’, allows for additional information to be included – rarely used</li>
</ul></li>
<li>Quality scores
<ul>
<li>Score for each base position in read – in ASCII code.</li>
</ul></li>
</ul>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastq-explainer1.png" class="img-fluid figure-img"></p>
<figcaption>fastq explainer</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="map-to-reference---sam-format" class="level2">
<h2 class="anchored" data-anchor-id="map-to-reference---sam-format">Map to reference - SAM format</h2>
<p>Each FASTQ contains thousands to millions of sequencing reads</p>
<p>Matching each read to the correct place in the genome (allowing for changes and errors) is computationally challenging</p>
<p>We use a ‘read aligner’ to map the position each read belongs to on a reference genome</p>
<ul>
<li>Short Reads - <code>BWA</code>, <code>Bowtie2</code></li>
<li>Long Reads - <code>minimap2</code></li>
</ul>
<p>Mapping produces a <code>SAM</code> file (or the compressed binary version <code>BAM</code> file)</p>
<p>We often use a tool called <code>samtools</code> to manage and view these large and complex files</p>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="SAM-explainer1.png" class="img-fluid figure-img"></p>
<figcaption>Samtool Image</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="variant-calling" class="level2">
<h2 class="anchored" data-anchor-id="variant-calling">Variant Calling</h2>
<p>When reads are aligned to a reference genome, point mutations (SNPs) between the sequencing reads and the reference can be identified. In the plot below: * Reads are in<span style="color:blue"> <em>blue</em> </span>/<span style="color:green"> <em>green</em></span>. * Changes from the reference are highlighted for each read in <span style="color:red"> <em>red</em></span>.</p>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="bam-ART-pileup.png" class="img-fluid figure-img"></p>
<figcaption>BAM-pileup</figcaption>
</figure>
</div>
<ul>
<li><p>Some variants can be very easy to determine However, sequencing error can lead to changes that are technical artefacts</p></li>
<li><p>Some variants are therefore more difficult to determine (e.g.&nbsp;a low coverage region with very few reads)</p></li>
<li><p>What if some reads correspond to a SNP and others do not?</p></li>
<li><p>Variant callers (e.g.&nbsp;<code>bcftools</code>, <code>GATK</code>, <code>FreeBayes</code>) apply statistical models to determine true variants, accounting for:</p></li>
<li><p>Reads counts/proportions supporting each allele</p></li>
<li><p>Quality scores for base, reads, mapping, position in the read a variant occurs</p></li>
<li><p>Variant callers output their analysis in a ‘variant call format’ – <code>VCF</code> file</p></li>
</ul>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="VCF.png" class="img-fluid figure-img"></p>
<figcaption>vcf</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="consensus-calling" class="level2">
<h2 class="anchored" data-anchor-id="consensus-calling">Consensus calling</h2>
<ul>
<li>Variant calls will be filtered and assessed (generally in an automated way).</li>
<li>Once we have a filtered variant set, variants can be applied to the reference genome.</li>
<li>This generates a consensus ‘pseudosequence’ in FASTA format.</li>
</ul>
<p><br></p>
<p><img src="consensus-fasta-Nhighlight.png" class="img-fluid" alt="consensus.fa"> <br></p>
<ul>
<li>Note, that one important step that can be overlooked is false negative variant calls
<ul>
<li>VCF format only highlights changes from the reference (i.e.&nbsp;positive calls)</li>
<li>If a base position has low coverage and no SNP is called, how do you know if there is a SNP or not?</li>
<li>Some methods will simply assume that the site is ‘reference’, without evidence to the contrary</li>
<li>This can give misleading results for some analyses (e.g.&nbsp;phylogenetics)</li>
<li>Important to ensure all sites without support are tested, and marked as ‘N’.</li>
</ul></li>
</ul>
<p><br></p>
<p>We can use integrated pipelines to automate parts of these processes</p>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Mapping_Overview_diag+pipelines.png" class="img-fluid figure-img"></p>
<figcaption>workflow-tools</figcaption>
</figure>
</div>
<p><br> <br></p>
</section>
</section>
<section id="practical-exercise---mapping-phylogeny" class="level1">
<h1>Practical Exercise - Mapping &amp; Phylogeny</h1>
<section id="finding-the-data" class="level2">
<h2 class="anchored" data-anchor-id="finding-the-data">Finding the data</h2>
<p>Navigate to the <code>Module_4_Mapping_Phylogeny</code> directory</p>
<pre class="text no-copy"><code>cd Module_4_Mapping_Phylogeny</code></pre>
<p>We can confirm where we are</p>
<pre><code>pwd</code></pre>
<p>We can also examine the contents of this file</p>
<pre><code>ls -l</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Directory_ls_start_2024.png" class="img-fluid figure-img"></p>
<figcaption>directory_contents</figcaption>
</figure>
</div>
<p><br></p>
<p>The folder contains:</p>
<ul>
<li>two pairs of sequencing reads :
<ul>
<li><code>new-sample-1_1.fastq.gz</code>, <code>new-sample-1_2.fastq.gz</code></li>
<li><code>new-sample-2_1.fastq.gz</code>, <code>new-sample-2_2.fastq.gz</code></li>
</ul></li>
<li>a folder containing indexed reference genomes</li>
<li>an archive file containing previous <code>snippy</code> runs (<code>snippy.runs.1.tar.gz</code>)</li>
</ul>
<p><br></p>
</section>
<section id="perform-mapping-of-reads-to-reference-genome-using-snippy" class="level2">
<h2 class="anchored" data-anchor-id="perform-mapping-of-reads-to-reference-genome-using-snippy">Perform mapping of reads to reference genome using snippy</h2>
<p><code>Snippy</code> is an integrated pipeline that maps reads to a reference genome and produces a range of outputs. For the purposes of this tutorial, we will use <code>snippy</code> to organise analyse our genomes.</p>
<p>You can view the options for snippy using the following code:</p>
<pre><code>snippy -h</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy_-h__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.h</figcaption>
</figure>
</div>
<p><br></p>
<p>Now run <code>snippy</code> on the pair of fastqs (for this exercise we’ll call this ‘new-sample-1’)</p>
<pre><code>snippy --outdir new-sample-1 --R1 new-sample-1_1.fastq.gz --R2 new-sample-1_2.fastq.gz --ref references/Vibrio_cholerae_O1_biovar_eltor_str_N16961_v2.fa --cpus 4 --ram 4 --force --quiet</code></pre>
<p>Wait for <code>snippy</code> to finish</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-run_1_2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1</figcaption>
</figure>
</div>
<p>Examine results of <code>snippy</code></p>
<pre><code>ls -lh new-sample-1/</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-pst_ls_2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.output</figcaption>
</figure>
</div>
<p>You can examine the log file to see exactly what <code>snippy</code> has done</p>
<pre><code>less new-sample-1/snps.log</code></pre>
<p><br></p>
<section id="examine-snippy-logs" class="level3">
<h3 class="anchored" data-anchor-id="examine-snippy-logs">Examine <code>snippy</code> logs</h3>
<p>We’ll use <code>grep</code> to retrieve the relevant lines for each command from the log file.</p>
<p>First, here is the command we used to set <code>snippy</code> running</p>
<pre><code>grep "outdir" new-sample-1/snps.log</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-grep-command__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.grep1</figcaption>
</figure>
</div>
<p>After ensuring the reference genome is indexed, <code>snippy</code> maps the reads to the reference genome using <code>bwa mem</code></p>
<pre><code>grep "bwa mem" new-sample-1/snps.log</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-bwa-command__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.grep2</figcaption>
</figure>
</div>
<p>Within that command, <code>snippy</code> also marks duplicate sequencing reads using <code>samtools markdup</code> (we extract only the first part of that command here, but you can look for it with <code>less</code>)</p>
<pre><code>grep "COMMAND: samtools" new-sample-1/snps.log</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-markdup__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.grep3</figcaption>
</figure>
</div>
<p><code>Snippy</code> then uses <code>FreeBayes</code> to call variants against the reference genome, producing a variant call file (.vcf)</p>
<pre><code>grep "freebayes" new-sample-1/snps.log</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-freebayes__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.grep4</figcaption>
</figure>
</div>
<p><code>Snippy</code> then applies some filters to assess the quality of those variants. It then applies the high quality variants to the reference genome to create a ‘pseudosequence consensus’ representation of our new genome</p>
<pre><code>grep "bcftools consensus" new-sample-1/snps.log</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-bcftools-consensus__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.grep5</figcaption>
</figure>
</div>
<p><code>Snippy</code> creates two versions of the pseudosequence consensus: - <code>snps.consensus.fa</code> contains all high quality variants - <code>snps.consensus.subs.fa</code> contains only high quality SNPs (no INDELS)</p>
<p><br></p>
</section>
<section id="examine-snippy-outputs" class="level3">
<h3 class="anchored" data-anchor-id="examine-snippy-outputs">Examine <code>snippy</code> outputs</h3>
<p>The <code>bam</code> file contains all the mapping positions on the genome for each individual read, along with metrics around mapping quality. This is a binary (machine readable) file, but we can view it using <code>samtools view</code></p>
<pre><code>samtools view new-sample-1/snps.bam | head -2</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-samtools-view-bam__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.samtools-view</figcaption>
</figure>
</div>
<p>The <code>VCF</code> file contains all the variants that have been called in our new genome compared to the reference genome</p>
<pre><code>head -35 new-sample-1/snps.vcf </code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-vcf__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.vcf</figcaption>
</figure>
</div>
<p>The first ~28 lines here are ‘headers’ and contain information about what has been done to call the variants, and helps you to interpret what different columns mean.</p>
<p>The last ~7 lines are individual variants, one per line. Variants columns are labelled</p>
<pre><code>#CHROM POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  new-sample</code></pre>
<p>We can view a slightly easier summary of these variants in the snps.tab file</p>
<pre><code>head -5 new-sample-1/snps.tab</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-snps.tab__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.snps</figcaption>
</figure>
</div>
<p>In this file, we have not provided gene information, so only the first 6 columns are relevant</p>
<p><code>Snippy</code> has also created a <code>fasta</code> file with our pseudosequence genome</p>
<pre><code>head new-sample-1/snps.consensus.fa</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-consensus.fasta.1__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy.run1.consensus.fa</figcaption>
</figure>
</div>
<p><br></p>
<p><br></p>
</section>
</section>
<section id="map-a-second-genome-to-reference" class="level2">
<h2 class="anchored" data-anchor-id="map-a-second-genome-to-reference">Map a second genome to reference</h2>
<p>We now need to map the second genome (<code>new-sample-2</code>) to our reference</p>
<p>Adapt the code examples provided above to run <code>snippy</code> on the second set of <code>fastq</code> files</p>
<p>Examine the results of snippy for <code>new-sample-2</code> using <code>grep</code>, <code>ls</code> and <code>less</code></p>
<p><br></p>
</section>
<section id="use-an-existing-assembly-as-input-to-snippy" class="level2">
<h2 class="anchored" data-anchor-id="use-an-existing-assembly-as-input-to-snippy">Use an existing assembly as input to snippy</h2>
<p>Sometimes we only have assemblies available, or we want to integrate long-read data into our analysis of short-read genomes. Snippy allows us to do this using the <code>--ctgs</code> argument. This takes an existing assembly file and simulates short reads, before calling variants against the reference.</p>
<p>We are going to integrate assemblies from a genome called “CTMA_1441” into our mapping analysis analysis. This genome was published in 2020 (https://pubmed.ncbi.nlm.nih.gov/32586863/), and used both short and long reads. For the purposes of this exercise, we have assembled it using three different approaches:</p>
<ul>
<li>Short read assembly (SPAdes) from Illumina reads</li>
<li>Long read assembly (Dragonflye) from Oxford Nanopore reads</li>
<li>Hybrid assembly from long reads (Dragonflye), followed by short read polishing</li>
</ul>
<p>Note that we will explore the differences between these assembly methods more in a later module.</p>
<p><br></p>
<section id="locate-existing-assemblies" class="level3">
<h3 class="anchored" data-anchor-id="locate-existing-assemblies">Locate existing assemblies</h3>
<p>The reads are hosted in the course GitHub repository. If that has been updated on your local VM, the files should be available to access. Let’s check:</p>
<pre><code>ls -lh ~/github_repository/course_data_2025/mapping_and_phylo/</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="github.repo.ls__2024.png" class="img-fluid figure-img"></p>
<figcaption>github.repo.ls</figcaption>
</figure>
</div>
<p>If the files are missing, check with the instructors.</p>
<p><br></p>
<p>Assuming they are available, we can copy the assemblies to our current working directory to make things easier:</p>
<pre><code>cp ~/github_repository/course_data_2025/mapping_and_phylo/CTMA_1441.* .</code></pre>
<p><br></p>
</section>
<section id="map-an-existing-short-read-assembly" class="level3">
<h3 class="anchored" data-anchor-id="map-an-existing-short-read-assembly">Map an existing short-read assembly</h3>
<p>The short read assembly is called <code>CTMA_1441.unicycler-short.fasta</code>. We can add the short-read assembly to snippy like this:</p>
<pre><code>snippy --outdir CTMA_1441.short --ctgs CTMA_1441.unicycler-short.fasta --ref references/Vibrio_cholerae_O1_biovar_eltor_str_N16961_v2.fa --cpus 4 --ram 4 --force --quiet</code></pre>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy_assembly_example__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy_assembly_example</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="map-a-long-read-assembly-to-reference" class="level3">
<h3 class="anchored" data-anchor-id="map-a-long-read-assembly-to-reference">Map a long-read assembly to reference</h3>
<p>We now need to map the two long read assemblies to our analysis: - <code>CTMA_1441.filtreads.dragonflye.long_4.1_polish.contigs.fa</code> - <code>CTMA_1441.filtreads.dragonflye.long_4.1_nopolish.contigs.fa</code></p>
<p><br></p>
<p>Adapt the code examples provided above to run <code>snippy</code> on the long-read assemblies using the output directories <code>CTMA_1441.long_polish</code> and <code>CTMA_1441.long_nopolish</code>.</p>
<p>Examine the results of snippy for each assembly using <code>grep</code>, <code>ls</code> and <code>less</code>.</p>
<p><br></p>
<p>What do you notice about the number of variable sites (reported at the end of the snippy run) identified with the different assemblies?</p>
<p><br></p>
</section>
</section>
<section id="contextualise-new-genomes-with-a-collection" class="level2">
<h2 class="anchored" data-anchor-id="contextualise-new-genomes-with-a-collection">Contextualise new genomes with a collection</h2>
<p>Now lets look at this new genome in context with a collection.</p>
<p>We can add this genome to a collection of genomes mapped using <code>snippy-core</code>.</p>
<p>For this exercise, we will again retrieve our data from the github folder:</p>
<pre><code>cp ~/github_repository/course_data_2025/mapping_and_phylo/old.snippy.runs_2025.tar.gz .</code></pre>
<p>Check the files are there:</p>
<pre><code>ls -lh</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-context.copy-git.ls__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-context.copy-git.ls</figcaption>
</figure>
</div>
<p>(Note that there is an existing file in our directory called `old.snippy.runs.2023.tar.gz - we will not be using this).</p>
<p><br></p>
<p>First, let’s extract the old <code>snippy</code> runs from our archive file</p>
<pre><code>tar -zxf old.snippy.runs_2025.tar.gz</code></pre>
<p>We can see that we now have a new directory <code>old.snippy.runs_2024</code></p>
<pre><code>ls -lh</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="untar.old-snippy-files.ls__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-context.tar.ls1</figcaption>
</figure>
</div>
<pre><code>ls -lh old.snippy.runs_2024</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="untar.old-snippy-files.ls2__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-context.tar.ls2</figcaption>
</figure>
</div>
<p><br></p>
<p>Now lets use <code>snippy-core</code> to summarise all these genomes along with the new ones and create a multiple sequence alignment</p>
<pre><code>snippy-core --ref references/Vibrio_cholerae_O1_biovar_eltor_str_N16961_v2.fa old.snippy.runs_2024/* new-sample-1 new-sample-2 CTMA_1441.short CTMA_1441.long_polish CTMA_1441.long_nopolish</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-core-run__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-core.run1</figcaption>
</figure>
</div>
<p><code>Snippy</code> has now created a number of files, including a ‘core SNP alignment’</p>
<pre><code>ls -l core.*</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-core.ls1__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-core.run2</figcaption>
</figure>
</div>
<p>We have various files that summarise our variants, e.g.</p>
<pre><code>head core.tab </code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-core_head-tab__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-core.run3</figcaption>
</figure>
</div>
<p>And our multiple sequence alignment containing all genomes:</p>
<pre><code>head core.full.aln</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-core_head.core.aln__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-core.run4</figcaption>
</figure>
</div>
<p>This file masks sequences with low confidence in different ways, but for some applications we want everything masked in the same way. Let’s change that so anything uncertain is marked as ’N’ using the <code>snippy-clean_full_aln</code> script that comes with <code>snippy</code>.</p>
<pre><code>snippy-clean_full_aln core.full.aln &gt; clean.full.aln</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snippy-core_clean.aln__2024.png" class="img-fluid figure-img"></p>
<figcaption>snippy-core.run5</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="phylogenetics" class="level2">
<h2 class="anchored" data-anchor-id="phylogenetics">Phylogenetics</h2>
<p>Now that we have a clean multiple sequence alignment, we are now going to use <code>IQ-TREE</code> to build a maximum likelihood phylogeny.</p>
<p><br></p>
<section id="make-a-snp-only-alignment-using-snp-sites" class="level3">
<h3 class="anchored" data-anchor-id="make-a-snp-only-alignment-using-snp-sites">Make a SNP-only alignment using <code>snp-sites</code></h3>
<p>Calculating a phylogeny on whole genome sequences can be very time consuming. We can speed this up by only using the variable sites (SNPs). However, we need to be aware that only including variable sites can affect the evolutionary rate estimates made by phylogenetics software - therefore, we need to account for the sites we remove in our analysis.</p>
<p>We will use <code>snp-sites</code> to do this. You can view the options for <code>snp-sites</code></p>
<pre><code>snp-sites</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snp-sites.help__2024.png" class="img-fluid figure-img"></p>
<figcaption>snp-sites.1</figcaption>
</figure>
</div>
<p>First, remove all the invariant sites and create a SNP-only multiple sequence alignment.</p>
<pre><code>snp-sites -o clean.full.SNPs.aln clean.full.aln</code></pre>
<p>We can also use snp-sites to find out how many invariant sites were removed (and what proportion of A, T, G, C they were). Use <code>snp-sites -C</code> to do this for the file above.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="snp-sites_run__2024.png" class="img-fluid figure-img"></p>
<figcaption>snp-sites.2</figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="calculate-a-phylogenetic-tree-from-the-snps-using-iq-tree2" class="level3">
<h3 class="anchored" data-anchor-id="calculate-a-phylogenetic-tree-from-the-snps-using-iq-tree2">Calculate a phylogenetic tree from the SNPs using <code>IQ-TREE2</code></h3>
<p>We can look at the options for <code>IQ-TREE</code> below</p>
<pre><code>iqtree -h</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="iqtree-help__2024.png" class="img-fluid figure-img"></p>
<figcaption>iqtree.1</figcaption>
</figure>
</div>
<p><br></p>
<p>In the command below, we: - specify the multiple sequence alignment using <code>-s clean.full.SNPs.aln</code> - ask <code>IQ-TREE</code> to take account of missing invariant sites using <code>-fconst $(snp-sites -C clean.full.aln)</code> - specify an evolutionary model we want <code>IQ-TREE</code> to use <code>-m GTR+F+I</code> - tell <code>IQ-TREE</code> to use a maximum of 2 CPUs (threads) and 2GB memory <code>-T 2 -mem 2G</code> - perform 1000 ultrafast bootstraps <code>-B 1000</code> - use sample <code>M66</code> as an outgroup <code>-o M66</code></p>
<pre><code>iqtree -s clean.full.SNPs.aln -fconst $( snp-sites -C clean.full.aln ) -m GTR+F+I -T 2 -mem 2G -B 1000 -o M66</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="iqtree-run.1__2024.png" class="img-fluid figure-img"></p>
<figcaption>iqtree.2</figcaption>
</figure>
</div>
<p><br></p>
<p>Look at folder</p>
<pre><code>ls -lh clean.full.*</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="iqtree_-ls-after__2024.png" class="img-fluid figure-img"></p>
<figcaption>iqtree.3</figcaption>
</figure>
</div>
<p><br></p>
<p>Our maximum likelihood tree is labelled <code>clean.full.SNPs.aln.treefile</code>. The <code>treefile</code> suffix is not always correctly identified by many tools, so we’ll relabel this as something else:</p>
<pre><code>cp clean.full.SNPs.aln.treefile clean.full.SNPs.aln.tre</code></pre>
<p><br></p>
</section>
<section id="visualise-a-phylogenetic-tree" class="level3">
<h3 class="anchored" data-anchor-id="visualise-a-phylogenetic-tree">Visualise a phylogenetic tree</h3>
<p>To examine our tree, we can look at the text file:</p>
<pre><code>cat clean.full.SNPs.aln.tre</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="iqtree_less-newick-file__2024.png" class="img-fluid figure-img"></p>
<figcaption>iqtree.4</figcaption>
</figure>
</div>
<p>But this is not very helpful - it’s just raw text in ‘newick’ format.</p>
<p>Instead, we can visualise this using <code>figtree</code></p>
<pre><code>figtree clean.full.SNPs.aln.tre &amp;</code></pre>
<p>(Note: Remember to include the <code>&amp;</code> symbol at the end - this allows figtree to run in the background, and allows you to still use the command prompt)</p>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figtree_launch__2024.png" class="img-fluid figure-img"></p>
<figcaption>figtree.1</figcaption>
</figure>
</div>
<p>Ignore the java error messages. The popup box is asking you how to describe the ‘bootstrap values’. You can click ‘OK’ here.</p>
<p>You should now have a visualisation of the tree we just generated</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figtree_initial.tree__2024.png" class="img-fluid figure-img"></p>
<figcaption>figtree.2</figcaption>
</figure>
</div>
<p><br></p>
<p>Examine the tree: - How are <code>new-sample-1</code> and <code>new-sample-2</code> related to each other? - What do you notice about the different versions of the assembled genomes <code>CTMA_1441</code>? - What does this tell you about the different assembly methods?</p>
<p><br> <br></p>
</section>
</section>
<section id="adjusting-a-dataset-to-remove-a-poor-quality-genome" class="level2">
<h2 class="anchored" data-anchor-id="adjusting-a-dataset-to-remove-a-poor-quality-genome">Adjusting a dataset to remove a poor quality genome</h2>
<p>The unpolished long-read assembly <code>CTMA_1441.long_nopolish</code> is filled with uncorrected errors. We therefore should not use it in our analysis. We will also remove the duplicate examples of CTMA_1441, so we are only using a single example. We can rerun <code>snippy-core</code> to rebuild the alignment without the unnecessary samples:</p>
<pre><code>snippy-core --ref references/Vibrio_cholerae_O1_biovar_eltor_str_N16961_v2.fa old.snippy.runs_2024/* new-sample-1 new-sample-2</code></pre>
<pre><code>snippy-clean_full_aln core.full.aln &gt; clean2.full.aln</code></pre>
<p><br></p>
<p>And now build a new phylogeny:</p>
<ul>
<li><p>Run <code>snp-sites</code> to create a file called <code>clean2.full.SNPs.aln</code></p></li>
<li><p>Run <code>iqtree</code> to calculate a phylogeny for <code>clean2.full.SNPs.aln</code></p></li>
<li><p>Rename the final treefile to <code>clean2.full.SNPs.aln.tre</code></p></li>
</ul>
<p><br></p>
<p>You can recheck the tree using <code>figtree</code>.</p>
<p><br> <br></p>
</section>
<section id="tree-visualisation-using-microreact" class="level2">
<h2 class="anchored" data-anchor-id="tree-visualisation-using-microreact">Tree Visualisation using Microreact</h2>
<p>We can also visualise our tree using a webtool called <code>Microreact</code></p>
<ul>
<li>Open Firefox and navigate to https://microreact.org/</li>
<li>Click ‘upload’</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Microreact-front-page.png" class="img-fluid figure-img"></p>
<figcaption>microreact.1</figcaption>
</figure>
</div>
<p><br></p>
<ul>
<li>Select the file upload, or drag the tree file <code>clean2.full.SNPs.aln.tre</code> onto the upload screen</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Microreact-upload.png" class="img-fluid figure-img"></p>
<figcaption>microreact.2</figcaption>
</figure>
</div>
<p><br></p>
<p>Microreact may not recognise the file, so let’s tell micoreact this this is a tree in <code>newick</code> format. Note that we can also use this screen to add metadata in <code>csv</code> format.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="microreact-specify-file-type.png" class="img-fluid figure-img"></p>
<figcaption>microreact.3</figcaption>
</figure>
</div>
<p><br></p>
<p>Click <code>continue</code> to view the tree</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="microreact-my-first-tree__2025.png" class="img-fluid figure-img"></p>
<figcaption>microreact.4</figcaption>
</figure>
</div>
<p><br> <br> <br></p>
</section>
<section id="accounting-for-recombination-with-gubbins" class="level2">
<h2 class="anchored" data-anchor-id="accounting-for-recombination-with-gubbins">Accounting for recombination with <code>gubbins</code></h2>
<p>We can use <code>gubbins</code> to infer recombining sites by looking for increased SNP density that occurs in specific ancestral nodes</p>
<p><code>gubbins</code> has been installed using <code>conda</code>. We must activate the relavent conda environment before running any commands using <code>gubbins</code>.</p>
<p><br></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate gubbins-env</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br></p>
<p>Now we can run <code>gubbins</code></p>
<pre><code>run_gubbins.py -h</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins-help__2024.png" class="img-fluid figure-img"></p>
<figcaption>gubbins.1</figcaption>
</figure>
</div>
<p><br></p>
<p>The following command runs <code>gubbins</code> on standard settings, with 4 CPUs.</p>
<p>Note: the <code>-c</code> option tells the program to use 4 CPUs. Note: the <code>-p</code> option tells the program to name all files with the prefix <code>gubbins</code> This command can take a few minutes to run.</p>
<pre><code>run_gubbins.py -c 4 -p gubbins clean2.full.aln</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins-run-hanging-screen__2024.png" class="img-fluid figure-img"></p>
<figcaption>run_gubbins_hanging</figcaption>
</figure>
</div>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins-run-finished-screen__2024.png" class="img-fluid figure-img"></p>
<figcaption>run_gubbins_hanging</figcaption>
</figure>
</div>
<p><br></p>
<p>Note: <code>gubbins</code> can take a long time to run on some computers. If <code>gubbins</code> takes more than 10 mins to complete, we have already run it for you - the files are available in the GitHub folder at <code>~/github_repository/Modules/Mapping_and_Phylo/gubbins_backup_2024.tar.gz</code>.</p>
<pre><code>ls -lh ~/github_repository/course_data_2025/mapping_and_phylo/</code></pre>
<pre><code>cp ~/github_repository/course_data_2025/mapping_and_phylo/gubbins_backup_2024.tar.gz .</code></pre>
<pre><code>tar -zxf gubbins_backup_2024.tar.gz</code></pre>
<p>Note: You may notice there is a directory called <code>gubbins_backups</code> - this is from a previous year, and will not be used today.</p>
<p><br> <br></p>
<p>Lets look at what gubbins has done</p>
<pre><code>ls -l gubbins.*</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins_backup_files__2024.png" class="img-fluid figure-img"></p>
<figcaption>gubbins.backup.files</figcaption>
</figure>
</div>
<p><br></p>
<p>You can explore these files For example <code>gubbins.recombination_predictions.gff</code> is a <code>gff</code> file that contains a record of each recombination block identified, how many SNPs it contains, and what samples are affected.</p>
<pre><code>head gubbins.recombination_predictions.gff</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins_view-gff__2024.png" class="img-fluid figure-img"></p>
<figcaption>gubbins.3</figcaption>
</figure>
</div>
<p><br></p>
<p><code>gubbins.filtered_polymorphic_sites.fasta</code> is a <code>fasta</code> file containing only SNPs, where recombining sites have been ‘masked to N’.</p>
<pre><code>head gubbins.filtered_polymorphic_sites.fasta</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins_polymorphic-sites__2024.png" class="img-fluid figure-img"></p>
<figcaption>gubbins.4</figcaption>
</figure>
</div>
<p><br></p>
<p><code>gubbins.final_tree.tre</code> is a phylogeny in which recombination has already been accounted for.</p>
<p>You can visualise this in <code>figtree</code> or <code>microreact</code> as above.<br>
<br></p>
<section id="root-a-phylogeny" class="level3">
<h3 class="anchored" data-anchor-id="root-a-phylogeny">Root a phylogeny</h3>
<p>The trees outputted by iqtree and gubbins are unrooted, but we may want to apply some evolutionary direction to them. One approach that is commonly used for bacterial datasets is midpoint rooting. Midpoint rooting involves locating the midpoint of the longest path between any two tips and putting the root in that location. Note that this does not necessarily infer the <em>true</em> root, and this should be used with caution.</p>
<p>To midpoint root our tree, we will use a simple script written in <code>python</code> that uses the <code>ete</code> package. You can examine the code:</p>
<pre><code>less midpoint.root.py</code></pre>
<p>To make the code work in the VM, we need to load another conda environment:</p>
<pre><code>conda activate ete3-env</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="midpoint.root.image.png" class="img-fluid figure-img"></p>
<figcaption>midpoint.script</figcaption>
</figure>
</div>
<p><br></p>
<p>We can use this script to midpoint root our tree</p>
<pre><code>python midpoint.root.py gubbins.final_tree.tre &gt; gubbins.final_tree.midpoint.tre</code></pre>
<p>You can visualise this in <code>figtree</code> or <code>microreact</code> as above. How does it compare to the unrooted version?</p>
<p><br> <br></p>
</section>
<section id="visualising-recombination-blocks-using-phandango" class="level3">
<h3 class="anchored" data-anchor-id="visualising-recombination-blocks-using-phandango">Visualising recombination blocks using phandango</h3>
<p>We can also visualise the tree together with the recombination blocks inferred by <code>gubbins</code> using a webtool called <code>phandango</code></p>
<p><br></p>
<p>Using your browser, navigate to (https://jameshadfield.github.io/phandango/#/)</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="phandango.landing.page.png" class="img-fluid figure-img"></p>
<figcaption>phandango.1</figcaption>
</figure>
</div>
<p><br></p>
<p>On your VM desktop, go to the file manager, and navigate to <code>/home/manager/Module_4_Mapping_Phylogeny</code></p>
<p>Drag and drop the tree file <code>gubbins.final_tree.midpoint.tre</code> into the phandango browser window</p>
<p>Drag and drop the recombination gff file <code>gubbins.recombination_predictions.gff</code> into the phandango browser window</p>
<p>Phandango should automatically display blocks of recombination in <span style="color:red"> <em>red</em> </span> (ancestral) and <span style="color:blue"> <em>blue</em> </span>(specific to a sample)</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gubbins-phandango-plot__2024.png" class="img-fluid figure-img"></p>
<figcaption>phandango.1</figcaption>
</figure>
</div>
<p><br> <br></p>
</section>
</section>
<section id="clustering-genomes-using-fastbaps" class="level2">
<h2 class="anchored" data-anchor-id="clustering-genomes-using-fastbaps">Clustering genomes using <code>fastBAPS</code></h2>
<p>We can cluster genomes for epidemiology in a variety of ways, depending on the goal and genetic distances involved. Here will use <code>fastBAPS</code>, which is an optimised implementation of the original <code>hierBAPS</code> algorithm for hierarchical partitioning and Bayesian clustering of genomes. <code>fastBAPS</code> can be run in <code>R</code> as well as from the command line.</p>
<p>In the command below, we: * specify a set of SNPs/alleles for use by the model. * <code>-i gubbins.filtered_polymorphic_sites.fasta</code> * specify the name of the output file. * <code>-o fastbaps.clusters.csv</code> * specify that we want hierarchical clustering at two different levels (i.e.&nbsp;BAPS will attempt to subdivide “level 1 clusters” to form “level 2 clusters”). * <code>--levels 2</code> * specify a model for delineating genomes - we will use a relaxed clustering model which should maximise the number of clusters we get at level 2. * <code>-p optimise.baps</code> * specify an existing phylogenetic tree that we want to use for conditioning the statistical model. This option is not essential, but it ensures that the tree and the clusters are consistent with one another. * <code>--phylogeny gubbins.final_tree.tre</code></p>
<pre><code>run_fastbaps -i gubbins.filtered_polymorphic_sites.fasta -o fastbaps.clusters --levels 2 -p optimise.baps --phylogeny gubbins.final_tree.tre</code></pre>
<p>We can view the output using <code>head</code> or <code>cat</code></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="run-fastbaps__2025.png" class="img-fluid figure-img"></p>
<figcaption>fastbaps.command</figcaption>
</figure>
</div>
<p><br></p>
<p>Now that we have clusters, let’s see how they look against the phylogenetic tree we made earlier.</p>
<p><br></p>
<p>First, we’ll need to make the <code>csv</code> file with our clusters compatible with <code>microreact</code>. We need to change the header of the first column from <code>Isolates</code> to <code>ID</code>. We’ll use some simple code on the command line to do this:</p>
<pre><code>sed s/Isolates/ID/ fastbaps.clusters &gt; fastbaps.clusters.fixid.csv</code></pre>
<pre><code>head fastbaps.clusters.fixid.csv</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="fastbaps.fixid__2025.png" class="img-fluid figure-img"></p>
<figcaption>fastbaps.fixid.code</figcaption>
</figure>
</div>
<p><br> <br></p>
<p>Go to the <code>microreact</code> webpage, and try uploading the new gubbins filtered tree and the fastbaps clusters</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="microreact.fastbaps.upload.png" class="img-fluid figure-img"></p>
<figcaption>microreact.fastbaps.upload</figcaption>
</figure>
</div>
<p><br></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="microreact.fastaps.upload2.png" class="img-fluid figure-img"></p>
<figcaption>microreact.fastbaps.upload2</figcaption>
</figure>
</div>
<p>You can see that the first column has been correctly identified as the ‘ID’ - this will be linked to the taxa in the tree</p>
<p>Click <code>Continue</code></p>
<p>You should now see a tree coloured by your fastbaps groups. You can use the toggles to change the colouring and to add metadata blocks.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="microreact.fastbaps.finaltree__2025.png" class="img-fluid figure-img"></p>
<figcaption>microreact.fastbaps.final</figcaption>
</figure>
</div>
<p><br></p>
<p><br> <br></p>
<p>END</p>
<p><a href="https://github.com/WCSCourses/GenEpiLAC2025/blob/main/course_modules_2025/Manual_main.md">&lt;&lt;&lt; Go back to Manual Contents Page</a></p>
<p><br></p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>This work is licensed under a Creative Commons Attribution 4.0 International License. Reuse is encouraged with acknowledgement</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>